scri/asymptotic_bondi_data/bms_charges.py

```diff
@@ -0,0 +1,223 @@
+# Copyright (c) 2020, Michael Boyle
+# See LICENSE file for details: <https://github.com/moble/scri/blob/master/LICENSE>
+
+### NOTE: The functions in this file are intended purely for inclusion in the AsymptoticBondData
+### class.  In particular, they assume that the first argument, `self` is an instance of
+### AsymptoticBondData.  They should probably not be used outside of that class.
+
+import numpy as np
+from math import sqrt, pi
+
+
+def mass_aspect(self, truncate_ell=max):
+    """Compute the Bondi mass aspect of the AsymptoticBondiData.
+
+    The Bondi mass aspect is given by
+
+        M = -ℜ{ψ₂ + σ ∂ₜσ̄}
+
+    Note that the last term is a product between two fields.  If, for example, these both have
+    ell_max=8, then their full product would have ell_max=16, meaning that we would go from
+    tracking 81 modes to 289.  This shows that deciding how to truncate the output ell is
+    important, which is why this function has the extra argument that it does.
+
+    Parameters
+    ==========
+    truncate_ell: int, or callable [defaults to `max`]
+        Determines how the ell_max value of the output is determined.  If an integer is passed,
+        each term in the output is truncated to have at most that ell_max.  (In particular,
+        terms that will not be used in the output are simply not computed, without incurring any
+        errors due to aliasing.)  If a callable is passed, it is passed on to the
+        spherical_functions.Modes.multiply method.  See that function's docstring for details.
+        The default behavior will result in the output having ell_max equal to the largest of
+        any of the individual Modes objects in the equation for M above -- but not the
+        product.
+
+    """
+    if callable(truncate_ell):
+        return -(self.psi2 + self.sigma.multiply(self.sigma.bar.dot, truncator=truncate_ell)).real
+    elif truncate_ell:
+        return -(
+            self.psi2.truncate_ell(truncate_ell)
+            + self.sigma.multiply(self.sigma.bar.dot, truncator=lambda tup: truncate_ell)
+        ).real
+    else:
+        return -(self.psi2 + self.sigma * self.sigma.bar.dot).real
+
+
+def charge_vector_from_aspect(charge):
+    """Output the ell<=1 modes of a BMS charge aspect as the charge four-vector."""
```

Comment:
Yup, this looks right to me. This also matches our equations in the remnants paper. I think `tᵝ + rᵝ` is more transparent for the follow reason:

As you've noted before, we are being a little bit loose with the word "aspect". In this file, we are taking "aspect" to be the integrand that gets projected along `lᵝ`, as you define here. This use of "aspect" correctly corresponds to the mass aspect when we are computing the Bondi four-momentum. So in your equation `vᵝ = (1/4π) ∫ ℜ{a} (tᵝ + rᵝ) dΩ`, our code and the literature would call `ℜ{a}` the mass aspect.

We start to stretch the term when it comes to the angular momentum and boost charges. For example, _in the literature_ the angular momentum aspect is often (ψ1+σðσ̄ + ½ð(σσ̄)) and it is projected not along `rᵝ` in the integral but along ð᷇rᵝ. So in this case the "literature angular momentum aspect" does not correspond to `ℜ{a}`. What we have done here is apply ð᷇ to rᵝ and work out the signs and factors so that our "aspect" `ℜ{a}` can be thrown into your equation  `vᵝ = (1/4π) ∫ ℜ{a} (tᵝ + rᵝ) dΩ` and we can get the vector. 

Because we often see the aspect projected along `lᵝ` in the literature, I think by writing ``tᵝ + rᵝ` instead it helps to signify that we are doing something a little bit different. I don't think any further comments are needed because you carefully define what we are calling the aspect within the doc string.

Reply:
Would you prefer that I add some further comments to the doc string to explain that a bit further?

Reply:
Nah, I think that's reasonable.